src/ray/raylet/node_manager.cc

```diff
@@ -1117,6 +1091,62 @@ void NodeManager::HandleWorkerUnblocked(std::shared_ptr<Worker> worker) {
   worker->MarkUnblocked();
 }
 
+void NodeManager::HandleClientBlocked(
+    const std::shared_ptr<LocalClientConnection> &client,
+    const std::vector<ObjectID> &required_object_ids) {
+  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+  if (worker) {
+    // The client is a worker. Mark the worker as blocked. This
+    // temporarily releases any resources that the worker holds while it is
+    // blocked.
+    HandleWorkerBlocked(worker);
+  } else {
+    // The client is a driver. Drivers do not hold resources, so we simply
+    // mark the driver as blocked.
+    worker = worker_pool_.GetRegisteredDriver(client);
+    RAY_CHECK(worker);
+    worker->MarkBlocked();
+  }
+  const TaskID current_task_id = worker->GetAssignedTaskId();
+  RAY_CHECK(!current_task_id.is_nil());
+  // Subscribe to the objects required by the ray.get. These objects will
+  // be fetched and/or reconstructed as necessary, until the objects become
+  // local or are unsubscribed.
+  task_dependency_manager_.SubscribeDependencies(current_task_id, required_object_ids);
```

Comment:
We should also unsubscribe the dependencies and cancel unfinished reconstruction after the timeout. 
Because if we're waiting for a very old object which takes long time to reconstruct, it'd be wasteful to continuing reconstructing it when wait already returns.

Reply:
Isn't that what we're doing at the bottom of `HandleClientUnblocked`? We're unsubscribing the dependencies. We're not canceling unfinished reconstruction I suppose, but it's not clear how to tell whether the reconstruction is needed or not. We could probably be less aggressive about triggering reconstruction from `ray.wait` (e.g., anytime there is a timeout, we don't really need to reconstruct anything), but I'm not sure if that's a good idea.